pkg/aws/blockindextablemapper.go

```diff
@@ -0,0 +1,160 @@
+}
+
+type BlockIndexRecord struct {
+	CarPath string
+	Offset  uint64
+	Length  uint64
+}
+
+type blockIndexTableMapper struct {
+	id              principal.Signer
+	blockIndexStore BlockIndexStore
+	bucketURL       url.URL
+}
+
+var _ providerindex.ContentToClaimsMapper = blockIndexTableMapper{}
+
+// NewBlockIndexTableMapper creates a new ContentToClaimsMapper that synthethizes location claims from data in the
+// blockIndexStore - a giant index of historical data, mapping multihashes to bucket keys/URLs and their byte offsets.
+//
+// The data referenced by bucket keys in the blockIndexStore has been consolidated into a single bucket. So this
+// instance does the work of mapping old bucket keys to URLs, where the base URL is the passed bucketURL param.
+//
+// Using the data in the blockIndexStore, the service will materialize content claims using the id param as the
+// signing key.
+func NewBlockIndexTableMapper(id principal.Signer, blockIndexStore BlockIndexStore, bucketURL string) (blockIndexTableMapper, error) {
+	burl, err := url.Parse(bucketURL)
+	if err != nil {
+		return blockIndexTableMapper{}, fmt.Errorf("parsing bucket URL: %w", err)
+	}
+
+	return blockIndexTableMapper{
+		id:              id,
+		blockIndexStore: blockIndexStore,
+		bucketURL:       *burl,
+	}, nil
+}
+
+// GetClaims implements providerindex.ContentToClaimsMapper.
+// Although it returns a list of CIDs, they are identity CIDs, so they contain the actual claims the refer to.
+func (bit blockIndexTableMapper) GetClaims(ctx context.Context, contentHash multihash.Multihash) ([]cid.Cid, error) {
+	var locs []cassert.LocationCaveats
+
+	// lets see if we can materialize some location claims
+	content := ctypes.FromHash(contentHash)
+	records, err := bit.blockIndexStore.Query(ctx, content.Hash())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, r := range records {
+		u, err := url.Parse(r.CarPath)
+		if err != nil || !u.IsAbs() {
+			// non-URL is legacy region/bucket/key format
+			// e.g. us-west-2/dotstorage-prod-1/raw/bafy...
+			parts := strings.Split(r.CarPath, "/")
+			key := strings.Join(parts[2:], "/")
+			shard, err := bucketKeyToShardLink(key)
+			if err != nil {
+				continue
+			}
+
+			u = bit.bucketURL.JoinPath(fmt.Sprintf("/%s/%s.car", shard.String(), shard.String()))
+			locs = append(locs, cassert.LocationCaveats{
+				Content:  content,
+				Location: []url.URL{*u},
+				Range:    &cassert.Range{Offset: r.Offset, Length: &r.Length},
+			})
+		} else {
+			locs = append(locs, cassert.LocationCaveats{
+				Content:  content,
+				Location: []url.URL{*u},
+				Range:    &cassert.Range{Offset: r.Offset, Length: &r.Length},
+			})
+		}
+	}
+
+	claimCids := make([]cid.Cid, 0, len(locs))
+	for _, loc := range locs {
+		claim, err := cassert.Location.Delegate(
+			bit.id,
+			bit.id,
+			bit.id.DID().String(),
+			loc,
+			delegation.WithExpiration(int(time.Now().Add(time.Hour).Unix())),
```

Comment:
I believe the redis cache will hold this for 1 hour also...we should probably make this slightly more than an hour so we don't return a cached but expired delegation later...

Reply:
Ideally we'd link these in code so a change to the cache expiration does not break this in the future.

Reply:
interesting. I linked synthetized claim expiration to the expiration in the cache. I added a buffer of 10 minutes, do you think that's a proper amount?

Reply:
um... not an  hour any more. try a month, less of a problem :P

Reply:
nm I now see  the design and have suggested what I think is the right value in the setup code.

Reply:
it's actually a week in prod now.